packages/yew-router/src/service.rs

```diff
@@ -1,90 +1,160 @@
-use crate::utils::base_url;
-use crate::Routable;
-use gloo::events::EventListener;
-use serde::{Deserialize, Serialize};
-use wasm_bindgen::JsValue;
-use web_sys::Event;
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use anymap::AnyMap;
 use yew::Callback;
 
-/// Navigate to a specific route.
-pub fn push_route(route: impl Routable) {
-    push_impl(route.to_path())
-}
+use super::history::{self, HistoryListener, Route};
+use crate::Routable;
 
-/// Navigate to a specific route with query parameters.
-///
-/// This should be used in cases where [`Link`](crate::prelude::Link) is insufficient.
-pub fn push_route_with_query<S>(
-    route: impl Routable,
-    query: S,
-) -> Result<(), serde_urlencoded::ser::Error>
-where
-    S: Serialize,
-{
-    let mut url = route.to_path();
-    let query = serde_urlencoded::to_string(query)?;
-    if !query.is_empty() {
-        url.push_str(&format!("?{}", query));
-    }
+type Entry<'a, T> = anymap::Entry<'a, dyn anymap::any::Any, T>;
 
-    push_impl(url);
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+enum RouterAction<T: Routable> {
+    Push(T),
+    Replace(T),
+}
+
+/// Specializes the history API for a particular `Routable` type
+pub(crate) struct RouterState<T: Routable> {
+    last_route: Rc<T>,
+    subscribers: Vec<Callback<Rc<T>>>,
+    _listener: HistoryListener,
+}
 
-    Ok(())
+thread_local! {
+    static ROUTER_STATE: RefCell<AnyMap> = RefCell::new(AnyMap::new());
```

Comment:
Why use `AnyMap`? It seems to be unmaintained (#1844) so shouldn't we use something else?

Besides, what's the purpose of this `thread_local`?

Reply:
You can't have generic static variables in Rust, so you need to use some kind of map to store the router state. We can use `anymap2` or one of the other crates that provides a similar API.

Reply:
I've replaced it with `anymap2` for now.

Reply:
Why do we even need this `thread_local`? I don't understand the problem it solves.

Reply:
That's what stores the RouterState.